dfusion_rust_core/src/models/order.rs

```diff
@@ -47,6 +47,90 @@ impl Order {
             sell_amount,
         }
     }
+
+    /// Calculates the utility of an order based on a buy and sell price as well
+    /// as the executed buy amount.
+    ///
+    /// # Returns
+    ///
+    /// Returns `Some(U256)` utility result if the inputs were valid and it
+    /// it produced a valid positive utility.
+    pub fn utility(
+        &self,
+        buy_price: u128,
+        sell_price: u128,
+        exec_buy_amount: u128,
+    ) -> Option<U256> {
+        // first lets convert everything into U256s
+        let buy_price = u128_to_u256(buy_price);
+        let sell_price = u128_to_u256(sell_price);
+        let exec_buy_amount = u128_to_u256(exec_buy_amount);
+        let buy_amount = u128_to_u256(self.buy_amount);
+        let sell_amount = u128_to_u256(self.sell_amount);
+        let fee_denominator = U256::from(2);
+
+        // adj_sell_amount =
+        //   (
+        //     (
+        //       (exec_buy_amount * buy_price) / (fee_denominator - 1)
+        //     ) *
+        //     fee_denominator
+        //   ) /
+        //   sell_price
+        //
+        // utility =
+        //   (
+        //     (exec_buy_amount * buy_price - adj_sell_amount * buy_amount) *
+        //     buy_price
+        //   ) /
+        //   sell_amount
+        let adj_sell_amount = exec_buy_amount
+            .checked_mul(buy_price)?
+            .checked_div(fee_denominator - 1)?
+            .checked_mul(fee_denominator)?
+            .checked_div(sell_price)?;
+        let utility = exec_buy_amount
+            .checked_mul(buy_price)?
+            .checked_sub(adj_sell_amount.checked_mul(buy_amount)?)?
+            .checked_mul(buy_price)?
+            .checked_div(sell_amount)?;
+
+        Some(utility)
```

Comment:
We must be cautious here that your utility is computed based on the formula in the contract's comment, but the contract actually makes one step of simplification canceling out `order.sell_amount` from the first term with the subtraction and moves the order of operations on the division of the "sell_amount". I believe this was done to reduce the amount of precision lost in the calculation and may have been missed because you currently have any extra copy of `buy_price` in the place where `order.sell_amount` should be!

Reply:
I think I just made a mistake copying the formula from the contract to my comment and then repeated it when implementing. I wasn't trying to be clever or anything.